### ISSUE TITLE
OCM CLI Usability Enhancements

### ISSUE THREAD

**Issue by phoban01** [open-component-model/ocm#144]
**What would you like to be added**:

If we wish to encourage adoption of the OCI specification and tooling there are some usability issues that we should address. The purpose here is to improve the developer experience using the OCM CLI.

The cumulative effect of the improvements outlined here will streamline the ocm cli workflow and should mean that it is not necessary to use a Makefile in order to execute a simple series of ocm commands. This will encourage adoption and enable straightforward integration into build processes and scripted workflows.

- [Component Initialization Process](component-initialization-process)
- [Target Directory](target-directory)
- [Target Parameter](target-parameter)
- [Target Parameter](target-parameter)
- [Creating Resources, Sources & References](creating-resources-sources-references)
- [Aliases & Inversions](aliases-inversions)
- [Unnamed Parameters](unnamed-parameters)
- [Examples](examples)
- [Surface important functionality](surface-important-functionality)

An example workflow with these changes incorporated could the as follows:

```bash!
# initialise our ocm component
$ ocm init github.com/torvalds/cloud-native-kernel --version v1.0.0

# examine the resulting "workspace" and component archive
$ tree
.
├── .ocm/
│   ├── blobs/
│   └── component-descriptor.yaml
    
# create a resource for the podinfo container image
$ ocm create resource podinfo \
    --type ociImage \
    --version 6.2.0 \
    --access-type ociArtefact \
    --access-image-reference ghcr.io/stefanprodan/podinfo:6.2.0

# sign the component version locally with the supplied key
$ ocm sign --local --key ocm-private.key --name devsig

# transfer the component to the github container registry and also move the podinfo resource
$ ocm transfer component --with-resources ghcr.io/linux

# verify after transfer
$ ocm verify --repository ghcr.io/linux github.com/torvalds/cloud-native-kernel:v1.0.0 

# sign the component version remotely with the supplied key
$ ocm sign --key private.key --name ownersig github.com/torvalds/cloud-native-kernel:v1.0.0
```

## Component Initialization Process

1  OCM initialisation workflow needs to be cleaner. An `ocm init` command should replace the `ocm create componentarchive` and be configurable with a single argument with options provided via named flags. 

The current command is as follows:
```bash
ocm create componentarchive github.com/phoban01/ocm-example v1.0.0 phoban01 out
```

The proposed `ocm init` command could be used as follows:

For example:
```bash
ocm init github.com/phoban01/ocm-example --version v1.0.0
```

Note that it is not necessary to specify a target directory and additional flags can supply information about the provider or scheme. If not present they can be defaulted and extracted from the name.

## Target Directory

The `ocm init` command should create an ocm content directory containing blobs and the component descriptor which is hidden from the user and doesn't pollute the working directory. By convention this could be called `.ocm`. This assumes the component archive working format would be a directory.

## Target Parameter

The `target` parameter should not be required for each command, use `.ocm` by convention or specify with --target flag

## Creating Resources, Sources & References

The ocm cli should enable generation of resources, sources or references. To create a new resource it should be possible to use a cli command rather than requiring to pass a configuration file:

```bash!
ocm create resource app-config --type TarArchive \
    --input-type file \
    --input-path ./resources/config.tar.gz \
    --input-media-type text/yaml.tar.gz
```

It should also be possible to export the yaml to stdout for further editing:
```bash!
ocm create resource app-config --type TarArchive \
    --input-type file \
    --input-path ./resources/config.tar.gz \
    --input-media-type text/yaml.tar.gz \
    --export > app-config.yaml
    
# ...edit app-config.yaml then add to component archive
ocm add resource app-config.yaml
```
## Aliases & Inversions

There are currently far too many aliases for CLI commands.

For example:
```bash!
ocm transfer componentversions | componentversion | cv | components | component | comps | comp | c
```
The `ocm transfer componentarchive` command can also be executed `ocm componentarchive transfer`. These types of inversions are unhelpful and confusing for users.

## Unnamed parameters

The CLI makes frequent use of unnamed positional paramaters. The CLI should bias towards a single unnamed parameter; multiple unnamed parameters should be discouraged and instead optional arguments should be accompanied by flags.

For example:

```bash!
ocm download resources <component> {<name>}
```

Should be:
```bash!
ocm download resources <component> {--name <name>}
```

## Examples

Many commands do not have a simple usage example. All should have a simple example and most should have more than one example.

## Surface important functionality

The CLI does lots of useful and interesting things but these are not apparent to the user.

For example, to transfer resource content alongside the component descriptor it is necessary to use the following syntax:

```bash!
 
ocm transfer component --resourcesByValue ghcr.io/phoban01//github.com/phoban01/wge-demo:v1.0.0 gcr.io/ocm-demo
```

This command should not have multiple aliases and should not obfuscate it's own utility using obscure flags.

A better alternative might be:
```bash!
ocm transfer component --with-resources ghcr.io/phoban01/wge-demo:v1.0.0 gcr.io/ocm-demo
```

Likewise the `ocm sign componentversion` and `ocm verify componentversion` commands could be simplified to remove unnecessary subcommands.

**Comment by In-Ko** [comment]
I'd like to pick this up again, as it seems very relevant that we simplify the usage of the OCM CLI as much as possible. Some of the proposed changes seem to be accepted, some not. I will try now to summarize what I have read above in the various comments.

> 1. [Uwe] The additional command inversion can be easily removed, no problem.

I fully support this. To many options on how to use a command are confusing.

> 2. Uwe] An init command as described would be possible, but it would require a fixed set of options, which are not optional.

Great, then let's introduce an `init` command with a fixed set of options for a default use case. This is also supported by Jason.

> 3. [Uwe] So, if you just prefer to use named arguments over positional ones this would be possible to change.

So wherever possible and meaningful, we will use named arguments instead of positional ones, with the one exception (brought up by Jens) if there are anyway only one or two arguments a command offers, we could still have positional arguments, simply as this required less typing and is still easy to understand. Jens and Jason also agree to this in their comments above.

> 4. Using a default folder or not

This point still seems to be discussed above without an agreed solution. Jason suggests the following above within the init command:

> Proposed way of doing it:
ocm init github.com/yitsushi/hello-world --version 1.0.7 --provider yitsushi
> or if  you want to have a specific directory
ocm init github.com/yitsushi/hello-world --version 1.0.7 --archivename ./hello-world-component

This makes sense to me.

---------

I hope we can discuss these changes briefly in one of the upcoming meetings (planning / Wednesday sync ...), agree on these 4 points and put them as individual small tasks on the backlog.

**Comment by mandelsoft** [comment]
- (1) The `create componentarchive` command has been simplified:
  - there is a default for the file name (`component-archive`) which ca be changed via option `--file`
  - Only two non-optional positional arguments, component name, and version 
- (2) the initial command basically is `create componentarchive`, I could add an additional alias `init`, but I think init alone does not directly imply what is initializes, a component, version, repository....
   *If preferred, I could rename the `create` command to `init`, but we should keep the object part, it would then be*
   ```
   ocm init componentarchive github.com/yitsushi/hello-world 1.0.7 --provider yitsushi
   ```

- The add commands have all been improved to be able to work without the external descriptions files purely based on options. But this is quite lengthy, because there are always lots of required settings . But it works.
- There are commands now to extend these descriptions files by command line options. (But this interferes with the possibility to use templating here, so it is only possible for non-templated entries)

Basically the orchestration commands are mainly intended for automated usage during a build process, for example as part of a makefile. Hereby, there is typically a predefined planned set of elements to add to a component archive.  For sure, the manual on-the fly usage is possible, but should be used as a playground, only.

**Comment by In-Ko** [comment]
As discussed in the review 2 weeks ago, this can be closed as a lot of the suggested improvements / changes have been done already.
